src/group_theory/monoid_localization.lean

```diff
@@ -266,25 +306,436 @@ show _ * _ = _, by rw [mul_inv_left, mul_one]
+
+/-- If `f : M →* N` is a localization map for a submonoid `S` and `k : N ≃* P` is an isomorphism
+    of `comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`. -/
+@[to_additive "If `f : M →+ N` is a localization map for a submonoid `S` and `k : N ≃+ P` is an isomorphism of `add_comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`."]
+def mul_equiv_map (k : N ≃* P) :
+  localization_map S P :=
+{ to_fun := k.to_monoid_hom.comp f.1,
+  map_units := λ y, is_unit_comp f k.to_monoid_hom y,
+  surj := λ v, let ⟨z, hz⟩ := k.to_equiv.surjective v in
+    let ⟨x, hx⟩ := f.3 z in ⟨x, show v * k _ = k _, by rw [←hx, k.map_mul, ←hz]; refl⟩,
+  eq_iff_exists := λ x y, (k.to_equiv.apply_eq_iff_eq _ _).trans $ f.4 _ _ }
+
+@[to_additive, simp] lemma mul_equiv_map_apply {k : N ≃* P} (x) :
+  (f.mul_equiv_map k).1 x = k (f.1 x) := rfl
+
+@[to_additive, simp] lemma mul_equiv_map_eq {k : N ≃* P} :
+  (f.mul_equiv_map k).1 = k.to_monoid_hom.comp f.1 := rfl
+
+@[to_additive] lemma symm_mul_equiv_map_apply {k : N ≃* P} (x) :
+  k.symm ((f.mul_equiv_map k).1 x) = f.1 x := k.symm_apply_apply (f.1 x)
+
+@[to_additive] lemma comp_mul_equiv_symm_map_apply {k : P ≃* N} (x) :
+  k ((f.mul_equiv_map k.symm).1 x) = f.1 x := k.apply_symm_apply (f.1 x)
+
+@[to_additive] lemma mul_equiv_map_eq_iff_eq {k : N ≃* P} {x y} :
+  (f.mul_equiv_map k).1 x = y ↔ f.1 x = k.symm y :=
+k.to_equiv.eq_symm_apply.symm
+
+@[to_additive] lemma mul_equiv_of_mul_equiv_map (k : localization_map S P) :
+  f.mul_equiv_map (f.mul_equiv_of_localizations k) = k :=
+to_fun_inj $ f.lift_comp k.2
+
+@[to_additive] lemma mul_equiv_of_mul_equiv_map_apply {k : localization_map S P} {x} :
+  (f.mul_equiv_map (f.mul_equiv_of_localizations k)).1 x = k.1 x :=
+ext_iff.1 (f.mul_equiv_of_mul_equiv_map k) x
+
+@[to_additive] lemma mul_equiv_map_of_mul_equiv (k : N ≃* P) :
+  f.mul_equiv_of_localizations (f.mul_equiv_map k) = k :=
+mul_equiv.ext $ monoid_hom.ext_iff.1 $ f.lift_of_comp k.to_monoid_hom
+
+@[to_additive] lemma mul_equiv_map_of_mul_equiv_apply {k : N ≃* P} (x) :
+  f.mul_equiv_of_localizations (f.mul_equiv_map k) x = k x :=
+by rw mul_equiv_map_of_mul_equiv
+
+@[simp, to_additive] lemma mul_equiv_map_id :
+  f.mul_equiv_map (mul_equiv.refl N) = f :=
+by ext; refl
+
+@[to_additive] lemma mul_equiv_map_comp {k : N ≃* P} {j : P ≃* Q} :
+  (f.mul_equiv_map (k.trans j)).1 = j.to_monoid_hom.comp (f.mul_equiv_map k).1 :=
+by ext; refl
+
+/-- Given `comm_monoid`s `M, P` and submonoids `S ⊆ M, T ⊆ P`, if `f : M →* N` is a localization
+    map for `S` and `k : P ≃* M` is an isomorphism of `comm_monoid`s such that `k(T) = S`, `f ∘ k`
+    is a localization map for `T`. -/
+@[to_additive "Given `comm_monoid`s `M, P` and submonoids `S ⊆ M, T ⊆ P`, if `f : M →* N` is a localization map for `S` and `k : P ≃* M` is an isomorphism of `comm_monoid`s such that `k(T) = S`, `f ∘ k` is a localization map for `T`."]
+def mul_equiv_comap {k : P ≃* M} (H : T.map k.to_monoid_hom = S) :
```

Comment:
To me the name `comap` means there's a contravariant functor somewhere. There might be a better name.